external-crates/move/crates/move-compiler/src/typing/core.rs

```diff
@@ -2162,3 +2278,119 @@ fn check_num_tvar_(subst: &Subst, ty: &Type) -> bool {
         _ => false,
     }
 }
+
+//--------------------------------------------------------------------------------------------------
+// Join for Ref Vars
+//--------------------------------------------------------------------------------------------------
+
+fn join_ref_var(
+    mut subst: Subst,
+    case: TypingCase,
+    loc1: Loc,
+    id1: RefVar,
+    lhs: &Type,
+    loc2: Loc,
+    id2: RefVar,
+    rhs: &Type,
+) -> Result<(Subst, Type), TypingError> {
+    use Type_::AutoRef;
+    let ref_var1 = forward_ref_var(&subst, id1);
+    let ref_var2 = forward_ref_var(&subst, id2);
+
+    let ref_kind1 = subst.get_ref_var(ref_var1);
+    let ref_kind2 = subst.get_ref_var(ref_var2);
+
+    let new_ref_var = RefVar::next();
+
+    match (ref_kind1, ref_kind2) {
+        (None, None) => {
+            // If both auto-refs are unset, combine them and join their types.
+            subst.insert_ref_var(ref_var1, RefKind::Forward(new_ref_var));
+            subst.insert_ref_var(ref_var2, RefKind::Forward(new_ref_var));
+            let (subst, ty) = join_impl(subst, case, lhs, rhs)?;
+            Ok((subst, sp(loc2, AutoRef(new_ref_var, Box::new(ty)))))
+        }
+        (Some(rk1), Some(rk2)) => {
+            assert!(!matches!(rk1, RefKind::Forward(_)));
+            assert!(!matches!(rk2, RefKind::Forward(_)));
+            if let Some(join_kind) = rk1.join(rk2) {
```

Comment:
I think this looks good! We just need to consider subtyping here

Reply:
i.e. we have `join(Imm, Mut) ==> Imm` but `subtype(Imm, Mut) ERROR`

Reply:
I changed this code so that the subtyping case won't update the refs, simply check if the first is a subtype of the second by realizing the refs. I believe that should be enough. If it's not, I will revisit it when I have code examples that will actually flow through that path.